src/librustc_mir/transform/generator.rs

```diff
@@ -0,0 +1,777 @@
+                &format!("generator contains type {} in MIR, but typeck only knows about {}",
+                    decl.ty,
+                    interior));
+        }
+    }
+
+    let upvar_len = mir.upvar_decls.len();
+    let live_decls : Vec<_> = mir.local_decls
+        .iter_enumerated_mut()
+        .filter(|&(local, _)| live_locals.contains(&local))
+        .collect();
+
+    let mut remap = HashMap::new();
+    let unit = tcx.mk_nil();
+    let mut vars: Vec<_> = live_decls.into_iter().enumerate().map(|(idx, (local, decl))| {
+        let var = decl.clone();
+        *decl = LocalDecl {
+            mutability: Mutability::Mut,
+            ty: unit,
+            name: None,
+            source_info,
+            internal: false,
+            is_user_variable: false,
+        };
+        remap.insert(local, (var.ty, upvar_len + 1 + idx));
+        var
+    }).collect();
+
+    let layout = GeneratorLayout {
+        fields: vars
+    };
+
+    (remap, layout)
+}
+
+fn insert_entry_point<'tcx>(mir: &mut Mir<'tcx>,
+                            block: BasicBlockData<'tcx>) {
+    mir.basic_blocks_mut().raw.insert(0, block);
+
+    let blocks = mir.basic_blocks_mut().iter_mut();
+
+    for target in blocks.flat_map(|b| b.terminator_mut().successors_mut()) {
+        *target = BasicBlock::new(target.index() + 1);
+    }
+}
+
+fn elaborate_generator_drops<'a, 'tcx>(tcx: TyCtxt<'a, 'tcx, 'tcx>,
+                                      def_id: DefId,
+                                      mir: &mut Mir<'tcx>) {
+    use util::elaborate_drops::{elaborate_drop, Unwind, DropElaborator, DropStyle, DropFlagMode};
+    use util::patch::MirPatch;
+    use shim::DropShimElaborator;
+
+    let param_env = tcx.param_env(def_id);
+    let gen = Local::new(1);
+
+    for block in mir.basic_blocks().indices() {
+        let (target, unwind, source_info) = match mir.basic_blocks()[block].terminator() {
+            &Terminator {
+                source_info,
+                kind: TerminatorKind::Drop {
+                    location: Lvalue::Local(local),
+                    target,
+                    unwind
+                }
+            } if local == gen => (target, unwind, source_info),
+            _ => continue,
+        };
+        let unwind = if let Some(unwind) = unwind {
+            Unwind::To(unwind)
+        } else {
+            Unwind::InCleanup
+        };
+        let patch = {
+            let mut elaborator = DropShimElaborator {
+                mir: &mir,
+                patch: MirPatch::new(mir),
+                tcx,
+                param_env
+            };
+            elaborate_drop(
+                &mut elaborator,
+                source_info,
+                &Lvalue::Local(gen),
+                (),
+                target,
+                unwind,
+                block
+            );
+            elaborator.patch
+        };
+        patch.apply(mir);
+    }
+}
+
+fn generate_drop<'a, 'tcx>(
```

Comment:
style: this should probably take the source mir by reference and return a cloned version of it by value.